--- a/package/openwrt/files/etc/init.d/smartdns
+++ b/package/openwrt/files/etc/init.d/smartdns
@@ -36,8 +36,8 @@ ADDRESS_CONF="$SMARTDNS_CONF_DIR/address.conf"
 BLACKLIST_IP_CONF="$SMARTDNS_CONF_DIR/blacklist-ip.conf"
 CUSTOM_CONF="$SMARTDNS_CONF_DIR/custom.conf"
 SMARTDNS_CONF_TMP="${SMARTDNS_CONF}.tmp"
-EXTRA_COMMANDS="updatefiles"
-EXTRA_HELP="        updatefiles      Update files"
+EXTRA_COMMANDS="updatefiles updatelist"
+EXTRA_HELP="        updatefiles      Update files      updatelist      update list"
 COREDUMP="0"
 RESPAWN="1"
 DO_RELOAD="0"
@@ -241,6 +241,35 @@ restart_crond()
 	/etc/init.d/cron restart >/dev/null 2>&1
 }
 
+disable_list_auto_update()
+{
+	local list_no_restart="$1"
+	grep "/usr/share/smartdns/update-list.sh updatelist" /etc/crontabs/root 1>/dev/null 2>&1
+	if [ $? -ne 0 ]; then
+		return 
+	fi
+
+	sed -i '\@/usr/share/smartdns/update-list.sh updatelist@d' /etc/crontabs/root
+
+	if [ "$list_no_restart" = "1" ]; then
+		return
+	fi
+
+	restart_crond
+}
+
+enable_list_auto_update()
+{
+	grep "$list_update_min_time $list_update_day_time * * $list_update_week_time /usr/share/smartdns/update-list.sh updatelist" /etc/crontabs/root 2>/dev/null
+	if [ $? -eq 0 ]; then
+		return 
+	fi
+
+	disable_list_auto_update 1
+	echo "$list_update_min_time $list_update_day_time * * $list_update_week_time /usr/share/smartdns/update-list.sh updatelist" >> /etc/crontabs/root
+	restart_crond
+}
+
 disable_auto_update()
 {
 	local no_restart="$1"
@@ -260,13 +289,13 @@ disable_auto_update()
 
 enable_auto_update()
 {
-	grep "0 5 * * * /etc/init.d/smartdns updatefiles" /etc/crontabs/root 2>/dev/null
+	grep "$auto_update_min_time $auto_update_day_time * * * /etc/init.d/smartdns updatefiles" /etc/crontabs/root 2>/dev/null
 	if [ $? -eq 0 ]; then
 		return 
 	fi
 
 	disable_auto_update 1
-	echo "0 5 * * * /etc/init.d/smartdns updatefiles" >> /etc/crontabs/root
+	echo "$auto_update_min_time $auto_update_day_time * * * /etc/init.d/smartdns updatefiles" >> /etc/crontabs/root
 	restart_crond
 }
 
@@ -498,6 +527,12 @@ load_service()
 	config_get tcp_server "$section" "tcp_server" "1"
 	config_get server_flags "$section" "server_flags" ""
 
+	config_get list_update_week_time "$section" "list_update_week_time" "*"
+	config_get list_update_day_time "$section" "list_update_day_time" "2"
+	config_get list_update_min_time "$section" "lsit_update_min_time" "30"
+	config_get auto_update_day_time "$section" "auto_update_day_time" "5"
+	config_get auto_update_min_time "$section" "auto_update_min_time" ""
+
 	config_get speed_check_mode "$section" "speed_check_mode" ""
 	[ ! -z "$speed_check_mode" ] && conf_append "speed-check-mode" "$speed_check_mode"
 
@@ -524,6 +559,9 @@ load_service()
 
 	config_get auto_set_dnsmasq "$section" "auto_set_dnsmasq" "1"
 
+	config_get enable_list_auto_update "$section" "enable_list_auto_update" "0"
+	[ "$enable_list_auto_update" = "1" ] && enable_list_auto_update || disable_list_auto_update
+
 	config_get ipset_no_speed "$section" "ipset_no_speed" ""
 	[ -z "$ipset_no_speed" ] || conf_append "ipset-no-speed" "$ipset_no_speed"
 
@@ -628,6 +666,7 @@ load_service()
 		[ "$old_enabled" = "0" ] && return 1
 		[ "$old_port" = "53" ] && stop_main_dns "0"
 		[ "$old_port" != "53" ] && [ "$old_auto_set_dnsmasq" = "1" ] && stop_forward_dnsmasq "$old_port" "0"
+		disable_list_auto_update
 		disable_auto_update
 		return 1
 	}
