diff --git a/Makefile b/Makefile
index f3ebac1..992d908 100644
--- a/Makefile
+++ b/Makefile
@@ -17,6 +17,8 @@ PKG_CONFIG_DEPENDS:= \
 	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_Haproxy \
 	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_Hysteria \
 	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_NaiveProxy \
+	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_PDNSD \
+	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_DNS2TCP \
 	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_Shadowsocks_Libev_Client \
 	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_Shadowsocks_Libev_Server \
 	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_Shadowsocks_Rust_Client \
@@ -35,13 +37,15 @@ PKG_CONFIG_DEPENDS:= \
 LUCI_TITLE:=LuCI support for PassWall
 LUCI_PKGARCH:=all
 LUCI_DEPENDS:=+coreutils +coreutils-base64 +coreutils-nohup +curl \
-	+dns2socks +dns2tcp +ip-full +libuci-lua +lua +luci-compat +luci-lib-jsonc \
+	+dns2socks +ip-full +libuci-lua +lua +luci-compat +luci-lib-jsonc \
 	+microsocks +resolveip +tcping +unzip \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_Brook:brook \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_ChinaDNS_NG:chinadns-ng \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_Haproxy:haproxy \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_Hysteria:hysteria \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_NaiveProxy:naiveproxy \
+	+PACKAGE_$(PKG_NAME)_INCLUDE_PDNSD:pdnsd-alt \
+	+PACKAGE_$(PKG_NAME)_INCLUDE_DNS2TCP:dns2tcp \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_Shadowsocks_Libev_Client:shadowsocks-libev-ss-local \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_Shadowsocks_Libev_Client:shadowsocks-libev-ss-redir \
 	+PACKAGE_$(PKG_NAME)_INCLUDE_Shadowsocks_Libev_Server:shadowsocks-libev-ss-server \
@@ -110,6 +114,14 @@ config PACKAGE_$(PKG_NAME)_INCLUDE_NaiveProxy
 	depends on !(arc||(arm&&TARGET_gemini)||armeb||mips||mips64||powerpc)
 	default n
 
+config PACKAGE_$(PKG_NAME)_INCLUDE_PDNSD
+	bool "Include PDNSD"
+	default y
+
+config PACKAGE_$(PKG_NAME)_INCLUDE_DNS2TCP
+	bool "Include DNS2TCP"
+	default y
+
 config PACKAGE_$(PKG_NAME)_INCLUDE_Shadowsocks_Libev_Client
 	bool "Include Shadowsocks Libev Client"
 	default y
diff --git a/luasrc/model/cbi/passwall/client/global.lua b/luasrc/model/cbi/passwall/client/global.lua
index d433b6d..9dbf384 100644
--- a/luasrc/model/cbi/passwall/client/global.lua
+++ b/luasrc/model/cbi/passwall/client/global.lua
@@ -343,8 +343,11 @@ end
 
 ---- DNS Forward Mode
 dns_mode = s:taboption("DNS", ListValue, "dns_mode", translate("Filter Mode"))
+if api.is_finded("pdnsd") then
+	dns_mode:value("pdnsd", "pdnsd " .. translatef("Requery DNS By %s", translate("TCP Node")))
+end
 if api.is_finded("dns2tcp") then
-	dns_mode:value("dns2tcp", translatef("Requery DNS By %s", "TCP"))
+	dns_mode:value("dns2tcp", "dns2tcp " ..translatef("Requery DNS By %s", "TCP")))
 end
 if api.is_finded("dns2socks") then
 	dns_mode:value("dns2socks", "dns2socks")
@@ -396,6 +399,7 @@ o:value("9.9.9.9", "9.9.9.9 (Quad9-Recommended)")
 o:value("208.67.220.220", "208.67.220.220 (OpenDNS)")
 o:value("208.67.222.222", "208.67.222.222 (OpenDNS)")
 o:depends({dns_mode = "dns2socks"})
+o:depends({dns_mode = "pdnsd"})
 o:depends({dns_mode = "dns2tcp"})
 o:depends({dns_mode = "udp"})
 o:depends({v2ray_dns_mode = "tcp"})
@@ -425,6 +429,8 @@ o:depends("v2ray_dns_mode", "doh")
 o = s:taboption("DNS", Flag, "dns_cache", translate("Cache Resolved"))
 o.default = "1"
 o:depends({dns_mode = "dns2socks"})
+o:depends({dns_mode = "pdnsd"})
+o:depends({dns_mode = "udp"})
 o:depends({dns_mode = "xray", v2ray_dns_mode = "tcp"})
 o:depends({dns_mode = "xray", v2ray_dns_mode = "doh"})
 o.rmempty = false
@@ -433,6 +439,7 @@ if api.is_finded("chinadns-ng") then
 	o = s:taboption("DNS", Flag, "chinadns_ng", translate("ChinaDNS-NG"), translate("The effect is better, but will increase the memory."))
 	o.default = "0"
 	o:depends({dns_mode = "dns2socks"})
+	o:depends({dns_mode = "pdnsd"})
 	o:depends({dns_mode = "dns2tcp"})
 	o:depends({dns_mode = "xray", v2ray_dns_mode = "tcp"})
 	o:depends({dns_mode = "xray", v2ray_dns_mode = "doh"})
diff --git a/luasrc/model/cbi/passwall/client/node_subscribe.lua b/luasrc/model/cbi/passwall/client/node_subscribe.lua
index 033a159..c3b2c56 100644
--- a/luasrc/model/cbi/passwall/client/node_subscribe.lua
+++ b/luasrc/model/cbi/passwall/client/node_subscribe.lua
@@ -103,7 +103,7 @@ o.validate = function(self, value, t)
 	end
 end
 
-o = s:option(DummyValue, "_node_count")
+o = s:option(DummyValue, "_node_count",translate("Node num"))
 o.rawhtml = true
 o.cfgvalue = function(t, n)
 	local remark = m:get(n, "remark") or ""
diff --git a/po/zh-cn/passwall.po b/po/zh-cn/passwall.po
index 2328aca..6343afb 100644
--- a/po/zh-cn/passwall.po
+++ b/po/zh-cn/passwall.po
@@ -241,6 +241,12 @@ msgstr "单进程"
 msgid "Proxy Mode"
 msgstr "代理模式"
 
+msgid "TCP Proxy Mode"
+msgstr "TCP 代理模式"
+
+msgid "UDP Proxy Mode"
+msgstr "UDP 代理模式"
+
 msgid "Default Proxy Mode"
 msgstr "默认代理模式"
 
diff --git a/root/usr/share/passwall/app.sh b/root/usr/share/passwall/app.sh
index 2cfb872..cd2ec5e 100755
--- a/root/usr/share/passwall/app.sh
+++ b/root/usr/share/passwall/app.sh
@@ -1174,6 +1174,12 @@ start_dns() {
 				run_v2ray ${_v2ray_args}
 			}
 		;;
+		pdnsd)
+			use_tcp_node_resolve_dns=1
+			gen_pdnsd_config "${dns_listen_port}" "${REMOTE_DNS}" "${DNS_CACHE}" "${DNS_MODE}"
+			ln_run "$(first_type pdnsd)" pdnsd "/dev/null" --daemon -c "${TMP_PATH}/pdnsd/pdnsd.conf" -d
+			echolog "  - 域名解析：pdnsd + 使用(TCP节点)解析域名..."
+		;;
 		dns2tcp)
 			use_tcp_node_resolve_dns=1
 			ln_run "$(first_type dns2tcp)" dns2tcp "/dev/null" -L "${TUN_DNS}" -R "$(get_first_dns REMOTE_DNS 53)" -v
@@ -1181,8 +1187,14 @@ start_dns() {
 		;;
 		udp)
 			use_udp_node_resolve_dns=1
+			if [ -f "$(first_type pdnsd)" ];then
+				gen_pdnsd_config "${dns_listen_port}" "${REMOTE_DNS}" "${DNS_CACHE}" "${DNS_MODE}"
+ 				ln_run "$(first_type pdnsd)" pdnsd "/dev/null" --daemon -c "${TMP_PATH}/pdnsd/pdnsd.conf" -d
+				echolog "  - 域名解析：pdnsd + 使用(UDP节点)解析域名..."
+			else
 			TUN_DNS="$(echo ${REMOTE_DNS} | sed 's/#/:/g' | sed -E 's/\:([^:]+)$/#\1/g')"
 			echolog "  - 域名解析：使用UDP协议请求DNS（$TUN_DNS）..."
+			fi
 		;;
 		esac
 
@@ -1239,6 +1251,74 @@ start_dns() {
 	esac
 }
 
+gen_pdnsd_config() {
+	local listen_port=${1}
+	local up_dns=${2}
+	local cache=${3}
+	local dns_mode=${4}
+	local pdnsd_dir=${TMP_PATH}/pdnsd
+	local perm_cache=2048
+	local _cache="on"
+	local query_method="tcp_only"
+	[ "$dns_mode" == "udp" ] && query_method="udp_only"
+
+	local reject_ipv6_dns=
+	[ "${cache}" = "0" ] && _cache="off" && perm_cache=0
+
+	mkdir -p "${pdnsd_dir}"
+	touch "${pdnsd_dir}/pdnsd.cache"
+	chown -R root.nogroup "${pdnsd_dir}"
+	if [ $PROXY_IPV6 == "0" ]; then
+		reject_ipv6_dns=$(cat <<- 'EOF'
+
+				reject = ::/0;
+				reject_policy = negate;
+		EOF
+		)
+	fi
+	cat > "${pdnsd_dir}/pdnsd.conf" <<-EOF
+		global {
+			perm_cache = $perm_cache;
+			cache_dir = "$pdnsd_dir";
+			run_as = "root";
+			server_ip = 127.0.0.1;
+			server_port = ${listen_port};
+			status_ctl = on;
+			query_method = ${query_method};
+			min_ttl = 1h;
+			max_ttl = 1w;
+			timeout = 10;
+			par_queries = 2;
+			neg_domain_pol = off;
+			udpbufsize = 1024;
+			proc_limit = 2;
+			procq_limit = 8;
+		}
+
+	EOF
+	echolog "  + [$?]Pdnsd (127.0.0.1:${listen_port})..."
+
+	append_pdnsd_updns() {
+		[ -z "${2}" ] && echolog "  | - 略过错误 : ${1}" && return 0
+		cat >> $pdnsd_dir/pdnsd.conf <<-EOF
+			server {
+				label = "node-${2}_${3}";
+				ip = ${2};
+				edns_query = on;
+				port = ${3};
+				timeout = 4;
+				interval = 10m;
+				uptest = none;
+				purge_cache = off;
+				proxy_only = on;
+				caching = $_cache;${reject_ipv6_dns}
+			}
+		EOF
+		echolog "  | - [$?]上游DNS：${2}:${3}"
+	}
+	hosts_foreach up_dns append_pdnsd_updns 53
+}
+
 add_ip2route() {
 	local ip=$(get_host_ip "ipv4" $1)
 	[ -z "$ip" ] && {
@@ -1620,7 +1700,7 @@ chnlist=$(echo "${TCP_PROXY_MODE}${LOCALHOST_TCP_PROXY_MODE}${UDP_PROXY_MODE}${L
 gfwlist=$(echo "${TCP_PROXY_MODE}${LOCALHOST_TCP_PROXY_MODE}${UDP_PROXY_MODE}${LOCALHOST_UDP_PROXY_MODE}" | grep "gfwlist")
 DNS_SHUNT=$(config_t_get global dns_shunt dnsmasq)
 [ -z "$(first_type $DNS_SHUNT)" ] && DNS_SHUNT="dnsmasq"
-DNS_MODE=$(config_t_get global dns_mode dns2tcp)
+DNS_MODE=$(config_t_get global dns_mode pdnsd)
 DNS_CACHE=$(config_t_get global dns_cache 0)
 REMOTE_DNS=$(config_t_get global remote_dns 1.1.1.1:53 | sed 's/#/:/g' | sed -E 's/\:([^:]+)$/#\1/g')
 CHINADNS_NG=$(config_t_get global chinadns_ng 0)
