diff --git a/luasrc/model/cbi/passwall/api/api.lua b/luasrc/model/cbi/passwall/api/api.lua
index 50d3e98..e6772e3 100755
--- a/luasrc/model/cbi/passwall/api/api.lua
+++ b/luasrc/model/cbi/passwall/api/api.lua
@@ -333,7 +333,7 @@ end
 
 function get_v2ray_version(file)
     if file == nil then file = get_v2ray_path() end
-    local cmd = "-version | awk '{print $2}' | sed -n 1P"
+    local cmd = "version | awk '{print $2}' | sed -n 1P"
     return get_bin_version_cache(file, cmd)
 end
 
@@ -614,4 +614,4 @@ function common_to_check(api_url, local_version, match_file_name)
         html_url = json.html_url,
         data = asset
     }
-end
\ No newline at end of file
+end
diff --git a/luasrc/model/cbi/passwall/server/api/app.lua b/luasrc/model/cbi/passwall/server/api/app.lua
index a61412f..274176e 100755
--- a/luasrc/model/cbi/passwall/server/api/app.lua
+++ b/luasrc/model/cbi/passwall/server/api/app.lua
@@ -123,7 +123,7 @@ local function start()
                 bin = ln_run("/usr/bin/" .. type .. "-server", type .. "-server", "-c " .. config_file .. " " .. udp_param, log_path)
             elseif type == "V2ray" then
                 config = require(require_dir .. "v2ray").gen_config(user)
-                bin = ln_run(api.get_v2ray_path(), "v2ray", "-config=" .. config_file, log_path)
+                bin = ln_run(api.get_v2ray_path(), "v2ray", "run " .. "-config=" .. config_file, log_path)
             elseif type == "Xray" then
                 config = require(require_dir .. "v2ray").gen_config(user)
                 bin = ln_run(api.get_xray_path(), "xray", "-config=" .. config_file, log_path)
@@ -194,4 +194,4 @@ if action then
 	elseif action == "stop" then
         stop()
 	end
-end
\ No newline at end of file
+end
diff --git a/root/usr/share/passwall/app.sh b/root/usr/share/passwall/app.sh
index d290963..f83e87c 100755
--- a/root/usr/share/passwall/app.sh
+++ b/root/usr/share/passwall/app.sh
@@ -351,7 +351,7 @@ run_v2ray() {
 	_extra_param="${_extra_param} -tcp_proxy_way $tcp_proxy_way"
 	_extra_param="${_extra_param} -loglevel $loglevel"
 	lua $API_GEN_V2RAY ${_extra_param} > $config_file
-	ln_run "$(first_type $(config_t_get global_app ${type}_file) ${type})" ${type} $log_file -config="$config_file"
+	ln_run "$(first_type $(config_t_get global_app ${type}_file) ${type})" ${type} $log_file run -config="$config_file"
 }
 
 run_dns2socks() {
