diff --git a/src/libs/common/Path.cpp b/src/libs/common/Path.cpp
index 0e162bb..713f6ee 100644
--- a/src/libs/common/Path.cpp
+++ b/src/libs/common/Path.cpp
@@ -230,6 +230,7 @@ CPath::CPath(const wxString& filename)
 	}
 
 	wxCharBuffer fn = filename2char(filename);
+	// add fn.length() for wx 3.x
 	if (fn.data()) {
 		// Filename is valid in the current locale. This means that
 		// it either originated from a (wx)system-call, or from a
