diff -urN aMule-2.1.3/amule.desktop aMule-2.1.3.new/amule.desktop
--- aMule-2.1.3/amule.desktop	2005-12-14 05:12:44.000000000 +0100
+++ aMule-2.1.3.new/amule.desktop	2006-06-17 13:52:51.000000000 +0200
diff --git a/amule.desktop b/amule.desktop
index aed90fd..3b014ca 100644
--- a/amule.desktop
+++ b/amule.desktop
@@ -1,10 +1,12 @@
 [Desktop Entry]
 Name=aMule
+Comment=Linux port of eMule client
+Comment[pl]=Linuksowy port klienta eMule
 Exec=amule %u
 Icon=amule
 Terminal=false
 Type=Application
-Categories=Network;P2P;
+Categories=GTK;Network;P2P;
 Comment=A client for the eD2k network
 Comment[fr]=Un client pour le réseau eD2k
 Comment[tr]=eD2k ağı için istemci
