--- a/src/Makefile
+++ b/src/Makefile
@@ -34,6 +34,7 @@ override CFLAGS += $(EXTRA_CFLAGS)
 ifdef VER
 override CFLAGS += -DSMARTDNS_VERION='"$(VER)"'
 endif
+override CFLAGS += -DFEATURE_CHECK_EDNS
 
 CXXFLAGS=-O2 -g -Wall -std=c++11 
 override CXXFLAGS +=-Iinclude
--- a/src/dns_server.c
+++ b/src/dns_server.c
@@ -5478,7 +5478,7 @@ static int _dns_server_do_query(struct d
 		}
 	}
 
-	ret = _dns_server_set_to_pending_list(request);
+	/* ret = _dns_server_set_to_pending_list(request); */
 	if (ret == 0) {
 		goto clean_exit;
 	}
