--- a/htdocs/luci-static/resources/view/smartdns/smartdns.js
+++ b/htdocs/luci-static/resources/view/smartdns/smartdns.js
@@ -166,6 +166,48 @@ return view.extend({
 		o.rmempty = false;
 		o.default = o.enabled;
 
+		// Update domain rules and IP list
+		o = s.taboption("settings", form.Flag, "enable_list_auto_update", _("Update domestic blacklist And whitelist"),
+			_("If required domestic ip segment whitelist,then in the custom settings Remove the # sign before 'conf-file'."));
+		o.rmempty = false;
+		o.default = o.enabled;
+
+		o = s.taboption("settings", form.ListValue, "list_update_week_time", _("Update Time (Every Week)"));
+		o.value('*', _('Every Day'));
+		o.value('1', _('Every Monday'));
+		o.value('2', _('Every Tuesday'));
+		o.value('3', _('Every Wednesday'));
+		o.value('4', _('Every Thursday'));
+		o.value('5', _('Every Friday'));
+		o.value('6', _('Every Saturday'));
+		o.value('0', _('Every Sunday'));
+		o.default = "*";
+		o.depends('enable_list_auto_update', '1');
+
+		o = s.taboption('settings', form.ListValue, 'list_update_day_time', _("Update time (every day)"));
+		for (var i = 0; i < 24; i++)
+			o.value(i, i + ':00');
+		o.default = '2';
+		o.depends('enable_list_auto_update', '1');
+
+		o = s.taboption('settings', form.ListValue, 'list_update_min_time', _("Update Interval (min)"));
+		for (var i = 0; i < 60; i++)
+			o.value(i, i + ':00');
+		o.default = '30';
+		o.depends('enable_list_auto_update', '1');
+
+		o = s.taboption('settings', form.DummyValue, "_list_update", _("Update List"));
+		o.renderWidget = function () {
+			return E('button', {
+				'class': 'btn cbi-button cbi-button-apply',
+				'id': 'btn_list_update',
+				'click': ui.createHandlerFn(this, function () {
+					return fs.exec('/usr/share/smartdns/update-list.sh')
+						.catch(function (e) { ui.addNotification(null, E('p', e.message), 'error') });
+				})
+			}, [_("Update")]);
+		}
+
 		///////////////////////////////////////
 		// advanced settings;
 		///////////////////////////////////////
@@ -563,6 +605,12 @@ return view.extend({
 		o.default = '5';
 		o.depends('enable_auto_update', '1');
 
+		o = s.taboption('files', form.ListValue, 'auto_update_min_time', _("Update Interval (min)"));
+		for (var i = 0; i < 60; i++)
+			o.value(i, i + ':00');
+		o.default = '';
+		o.depends('enable_auto_update', '1');
+
 		o = s.taboption("files", form.FileUpload, "upload_conf_file", _("Upload Config File"),
 			_("Upload smartdns config file to /etc/smartdns/conf.d"));
 		o.rmempty = true
@@ -726,6 +774,7 @@ return view.extend({
 		o.depends("type", "udp");
 		o.depends("type", "tcp");
 		o.depends("type", "tls");
+		o.depends("type", "https");
 
 		// type;
 		o = s.taboption("general", form.ListValue, "type", _("type"), _("DNS Server type"));
--- a/po/zh_Hans/smartdns.po
+++ b/po/zh_Hans/smartdns.po
@@ -61,6 +61,14 @@ msgstr "自动设置Dnsmasq"
 msgid "Automatically set as upstream of dnsmasq when port changes."
 msgstr "端口更改时自动设为 dnsmasq 的上游。"
 
+#: applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js:170
+msgid "Update domestic blacklist And whitelist"
+msgstr "自动更新国内黑白名单"
+
+#: applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js:171
+msgid "If required domestic ip segment whitelist,then in the custom settings Remove the # sign before 'conf-file'."
+msgstr "如需使用国内 IP 段白名单，则在自定义设置中去除 conf-file 前面的 # 号。"
+
 #: applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js:1269
 msgid "Blacklist IP"
 msgstr "黑名单"
--- a/root/usr/share/rpcd/acl.d/luci-app-smartdns.json
+++ b/root/usr/share/rpcd/acl.d/luci-app-smartdns.json
@@ -14,7 +14,8 @@
 			"file": {
 				"/etc/smartdns/*": [ "write" ],
 				"/etc/init.d/smartdns restart": [ "exec" ],
-				"/etc/init.d/smartdns updatefiles": [ "exec" ]
+				"/etc/init.d/smartdns updatefiles": [ "exec" ],
+				"/usr/share/smartdns/update-list.sh": [ "exec" ]
 			},
 			"uci": [ "smartdns" ]
 		}
