--- a/htdocs/luci-static/resources/view/smartdns/smartdns.js
+++ b/htdocs/luci-static/resources/view/smartdns/smartdns.js
@@ -162,6 +162,48 @@ return view.extend({
 		o.rmempty = false;
 		o.default = o.enabled;
 
+		// Update domain rules and IP list
+		o = s.taboption("settings", form.Flag, "enable_list_auto_update", _("Update Domain rule list And Adblock Database"),
+			_("If required domain rules,then in domain name rule list configure.If required Adblock,then in the custom settings Remove the # sign before 'conf-file'."));
+		o.rmempty = false;
+		o.default = o.enabled;
+
+		o = s.taboption("settings", form.ListValue, "list_update_week_time", _("Update Time (Every Week)"));
+		o.value('*', _('Every Day'));
+		o.value('1', _('Every Monday'));
+		o.value('2', _('Every Tuesday'));
+		o.value('3', _('Every Wednesday'));
+		o.value('4', _('Every Thursday'));
+		o.value('5', _('Every Friday'));
+		o.value('6', _('Every Saturday'));
+		o.value('0', _('Every Sunday'));
+		o.default = "*";
+		o.depends('enable_list_auto_update', '1');
+
+		o = s.taboption('settings', form.ListValue, 'list_update_day_time', _("Update time (every day)"));
+		for (var i = 0; i < 24; i++)
+			o.value(i, i + ':00');
+		o.default = '2';
+		o.depends('enable_list_auto_update', '1');
+
+		o = s.taboption('settings', form.ListValue, 'list_update_min_time', _("Update Interval (min)"));
+		for (var i = 0; i < 60; i++)
+			o.value(i, i + ':00');
+		o.default = '30';
+		o.depends('enable_list_auto_update', '1');
+
+		o = s.taboption('settings', form.DummyValue, "_list_update", _("Update List"));
+		o.renderWidget = function () {
+			return E('button', {
+				'class': 'btn cbi-button cbi-button-apply',
+				'id': 'btn_list_update',
+				'click': ui.createHandlerFn(this, function () {
+					return fs.exec('/usr/share/smartdns/update-list.sh')
+						.catch(function (e) { ui.addNotification(null, E('p', e.message), 'error') });
+				})
+			}, [_("Update")]);
+		}
+
 		///////////////////////////////////////
 		// advanced settings;
 		///////////////////////////////////////
@@ -497,6 +539,18 @@ return view.extend({
 		o.default = o.disabled;
 		o.rempty = true;
 
+		o = s.taboption('files', form.ListValue, 'auto_update_day_time', _("Update time (every day)"));
+		for (var i = 0; i < 24; i++)
+			o.value(i, i + ':00');
+		o.default = '5';
+		o.depends('enable_auto_update', '1');
+
+		o = s.taboption('files', form.ListValue, 'auto_update_min_time', _("Update Interval (min)"));
+		for (var i = 0; i < 60; i++)
+			o.value(i, i + ':00');
+		o.default = '';
+		o.depends('enable_auto_update', '1');
+
 		o = s.taboption("files", form.FileUpload, "upload_conf_file", _("Upload Config File"),
 			_("Upload smartdns config file to /etc/smartdns/conf.d"));
 		o.rmempty = true
@@ -660,6 +714,7 @@ return view.extend({
 		o.depends("type", "udp");
 		o.depends("type", "tcp");
 		o.depends("type", "tls");
+		o.depends("type", "https");
 
 		// type;
 		o = s.taboption("general", form.ListValue, "type", _("type"), _("DNS Server type"));
--- a/po/zh_Hans/smartdns.po
+++ b/po/zh_Hans/smartdns.po
@@ -56,6 +56,48 @@ msgstr "自动设置Dnsmasq"
 msgid "Automatically set as upstream of dnsmasq when port changes."
 msgstr "端口更改时自动设为 dnsmasq 的上游。"
 
+msgid "Update Domain rule list And Adblock Database"
+msgstr "自动更新域名规则和广告数据库"
+
+msgid "If required domain rules,then in domain name rule list configure.If required Adblock,then in the custom settings Remove the # sign before 'conf-file'."
+msgstr "如需使用域名规则，则在域名规则列表配置。如需广告屏蔽，则在自定义设置中去除 conf-file 前面的 # 号。"
+
+msgid "Update Time (Every Week)"
+msgstr "更新时间(每周)"
+
+msgid "Every Day"
+msgstr "每天"
+
+msgid "Every Monday"
+msgstr "每周一"
+
+msgid "Every Tuesday"
+msgstr "每周二"
+
+msgid "Every Wednesday"
+msgstr "每周三"
+
+msgid "Every Thursday"
+msgstr "每周四"
+
+msgid "Every Friday"
+msgstr "每周五"
+
+msgid "Every Saturday"
+msgstr "每周六"
+
+msgid "Every Sunday"
+msgstr "每周日"
+
+msgid "Update time (every day)"
+msgstr "更新时间(每天)"
+
+msgid "Update Interval (min)"
+msgstr "更新间隔(分钟)"
+
+msgid "Update List"
+msgstr "更新列表"
+
 #: applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js:229
 msgid "Bind Device"
 msgstr "绑定到设备"
@@ -402,7 +444,7 @@ msgid ""
 "downloaded from the download page."
 msgstr ""
 "包含配置文件，路径为/etc/smartdns/conf.d，或自定义配置文件路径，可以从下载页"
-"配置自动下载。面配置自动下载。"
+"配置自动下载。"
 
 #: applications/luci-app-smartdns/htdocs/luci-static/resources/view/smartdns/smartdns.js:283
 msgid "Ipset name, Add domain result to ipset when speed check fails."
--- a/root/usr/share/rpcd/acl.d/luci-app-smartdns.json
+++ b/root/usr/share/rpcd/acl.d/luci-app-smartdns.json
@@ -14,7 +14,8 @@
 			"file": {
 				"/etc/smartdns/*": [ "write" ],
 				"/etc/init.d/smartdns restart": [ "exec" ],
-				"/etc/init.d/smartdns updatefiles": [ "exec" ]
+				"/etc/init.d/smartdns updatefiles": [ "exec" ],
+				"/usr/share/smartdns/update-list.sh": [ "exec" ]
 			},
 			"uci": [ "smartdns" ]
 		}
