--- a/luasrc/model/cbi/openclash/config-subscribe-edit.lua
+++ b/luasrc/model/cbi/openclash/config-subscribe-edit.lua
@@ -105,6 +105,9 @@
 o:value("https://api.dler.io/sub", translate("api.dler.io")..translate("(Default)"))
 o:value("https://api.wcc.best/sub", translate("api.wcc.best"))
 o:value("https://api.asailor.org/sub", translate("api.asailor.org"))
+o:value("https://api.sub.x10.bz/sub", translate("api.sub.x10.bz")..translate("(Support Vless By Oneself)"))
+o:value("https://api.api88.cloudns.org/sub", translate("api.api88.cloudns.org")..translate("(Support Vless By Oneself)"))
+o:value("https://api.api88.cloudns.be/sub", translate("api.api88.cloudns.be")..translate("(Support Vless By Cloudflare)"))
 o.default = "https://api.dler.io/sub"
 o.placeholder = "https://api.dler.io/sub"
 
--- a/po/zh-cn/openclash.zh-cn.po
+++ b/po/zh-cn/openclash.zh-cn.po
@@ -2844,6 +2844,12 @@ msgstr "启用后将尽可能的覆盖 DNS 查询到的域名"
 msgid "Override All Dns Query"
 msgstr "启用后将尽可能的覆盖 DNS 查询到的域名"
 
+msgid "(Support Vless By Oneself)"
+msgstr "(自建项目，支持 Vless)"
+
+msgid "(Support Vless By Cloudflare)"
+msgstr "(Cloudflare 项目，支持 Vless)"
+
 msgid "Meta Settings"
 msgstr "Meta 设置"
 
--- /dev/null
+++ b/po/zh_Hans
@@ -0,0 +1 @@
+zh-cn
\ No newline at end of file
