--- a/Makefile
+++ b/Makefile
@@ -9,7 +9,28 @@ PKG_BUILD_DIR:=$(BUILD_DIR)/$(PKG_NAME)
 
 include $(INCLUDE_DIR)/package.mk
 
+PKG_CONFIG_DEPENDS:= \
+	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_DEV \
+	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_PREMIUM \
+	CONFIG_PACKAGE_$(PKG_NAME)_INCLUDE_META
+
 define Package/$(PKG_NAME)/config
+menu "Configuration"
+
+config PACKAGE_$(PKG_NAME)_INCLUDE_DEV
+	bool "Include OPENCLASH-DEV"
+	default n
+
+config PACKAGE_$(PKG_NAME)_INCLUDE_PREMIUM
+	bool "Include OPENCLASH-PREMIUM"
+	default n
+
+config PACKAGE_$(PKG_NAME)_INCLUDE_META
+	bool "Include OPENCLASH-META"
+	default n
+
+endmenu
+
 	config PACKAGE_kmod-inet-diag
 	default y if PACKAGE_$(PKG_NAME)
 
@@ -34,8 +55,12 @@ define Package/$(PKG_NAME)
 	SUBMENU:=3. Applications
 	TITLE:=LuCI support for clash
 	PKGARCH:=all
-	DEPENDS:=+dnsmasq-full +coreutils +coreutils-nohup +bash +curl +ca-certificates +ipset +ip-full \
-	+libcap +libcap-bin +ruby +ruby-yaml +kmod-tun +unzip
+	DEPENDS:= \
+		+dnsmasq-full +coreutils +coreutils-nohup +bash +curl +ca-certificates +ipset +ip-full \
+		+libcap +libcap-bin +ruby +ruby-yaml +kmod-tun +unzip \
+		+PACKAGE_$(PKG_NAME)_INCLUDE_DEV:openclash-dev \
+		+PACKAGE_$(PKG_NAME)_INCLUDE_PREMIUM:openclash-premium \
+		+PACKAGE_$(PKG_NAME)_INCLUDE_META:openclash-meta
 	MAINTAINER:=vernesong
 endef
 
--- a/luasrc/model/cbi/openclash/config-subscribe-edit.lua
+++ b/luasrc/model/cbi/openclash/config-subscribe-edit.lua
@@ -105,6 +105,8 @@ o:value("https://api.dler.io/sub", translate("api.dler.io")..translate("(Default
 o:value("https://v.id9.cc/sub", translate("v.id9.cc")..translate("(Support Vless By Pinyun)"))
 o:value("https://sub.id9.cc/sub", translate("sub.id9.cc"))
 o:value("https://api.wcc.best/sub", translate("api.wcc.best"))
+o:value("https://api.api88.cloudns.org/sub", translate("api.api88.cloudns.org")..translate("(Support Vless)"))
+o:value("https://api.api88.cloudns.be/sub", translate("api.api88.cloudns.be")..translate("(Support Vless By Cloudflare)"))
 o.default = "https://api.dler.io/sub"
 o.placeholder = "https://api.dler.io/sub"
 
--- a/luasrc/model/cbi/openclash/settings.lua
+++ b/luasrc/model/cbi/openclash/settings.lua
@@ -38,7 +38,7 @@ m.description = translate("Note: To restore the default configuration, try acces
 "<br/>"..font_green..translate("Note: Game proxy please use nodes except VMess")..font_off..
 "<br/>"..font_green..translate("Note: If you need to perform client access control in Fake-IP mode, please change the DNS hijacking mode to firewall forwarding")..font_off..
 "<br/>"..translate("Note: The default proxy routes local traffic, BT, PT download, etc., please use Redir-Host mode as much as possible and pay attention to traffic avoidance")..
-"<br/>"..translate("Note: If the connection is abnormal, please follow the steps on this page to check first")..": ".."<a href='javascript:void(0)' onclick='javascript:return winOpen(\"https://github.com/vernesong/OpenClash/wiki/%E7%BD%91%E7%BB%9C%E8%BF%9E%E6%8E%A5%E5%BC%82%E5%B8%B8%E6%97%B6%E6%8E%92%E6%9F%A5%E5%8E%9F%E5%9B%A0\")'>"..translate("Click to the page").."</a>"
+"<br/>"..translate("Note: If the connection is abnormal, please follow the steps on this page to check first")..": ".."<a href='javascript:void(0)' onclick='javascript:return winOpen(\"https://github.com/vernesong/OpenClash/wiki/%E7%BD%91%E7%BB%9C%E8%BF%9E%E6%8E%A5%E5%BC%82%E5%B8%B8%E6%97%B6%E6%8E%92%E6%9F%A5%E5%8E%9F%E5%9B%A0\")'>"..font_green..bold_on..translate("Click to the page")..bold_off..font_off.."</a>"
 
 s = m:section(TypedSection, "openclash")
 s.anonymous = true
--- a/po/zh-cn/openclash.zh-cn.po
+++ b/po/zh-cn/openclash.zh-cn.po
@@ -2970,6 +2970,12 @@ msgstr "启用后将尽可能的覆盖DNS查询到的域名"
 msgid "(Support Vless By Pinyun)"
 msgstr "(由品云提供，支持 Vless)"
 
+msgid "(Support Vless)"
+msgstr "(支持 Vless)"
+
+msgid "(Support Vless By Cloudflare)"
+msgstr "(Cloudflare 项目，支持 Vless)"
+
 msgid "Meta Settings"
 msgstr "Meta 设置"
 
--- /dev/null
+++ b/po/zh_Hans
@@ -0,0 +1 @@
+zh-cn
\ No newline at end of file
