--- a/easyrsa3/easyrsa
+++ b/easyrsa3/easyrsa
@@ -5643,8 +5643,17 @@ select_vars() {
 			fi
 		fi
 
+		# User specified "$PWD/pki"; if vars exists, use it ONLY
+		if [ "$PWD/pki" ]; then
+			if [ -e "$PWD/pki/vars" ]; then
+				verbose "select_vars: PWD/pki/vars"
+				set_var EASYRSA_VARS_FILE "$PWD/pki/vars"
+			fi
+		fi
+
 		# Default PKI; if vars exists, use it ONLY
-		if [ -e "$PWD/pki/vars" ] && \
+		if [ -e "/etc/easy-rsa/vars" ] && \
+			[ -z "$PWD/pki" ] && \
 			[ -z "$EASYRSA_PKI" ] && \
 			[ -z "$EASYRSA" ]
 		then
@@ -5663,14 +5672,14 @@ select_vars() {
 			#set_var EASYRSA "$PWD"
 			#set_var EASYRSA_PKI "$EASYRSA/pki"
 
-			verbose "select_vars: PWD/pki/vars"
-			set_var EASYRSA_VARS_FILE "$PWD/pki/vars"
+			verbose "select_vars: /etc/easy-rsa/vars"
+			set_var EASYRSA_VARS_FILE "/etc/easy-rsa/vars"
 		fi
 
 		# Default working dir; if vars exists, use it ONLY
-		if [ -e "$PWD/vars" ]; then
-			verbose "select_vars: PWD/vars"
-			set_var EASYRSA_VARS_FILE "$PWD/vars"
+		if [ -e "/etc/easy-rsa/vars" ]; then
+			verbose "select_vars: /etc/easy-rsa/vars"
+			set_var EASYRSA_VARS_FILE "/etc/easy-rsa/vars"
 		fi
 	fi
 
@@ -5773,7 +5782,7 @@ default_vars() {
 	# Set defaults, preferring existing env-vars if present
 	set_var EASYRSA					"$PWD"
 	set_var EASYRSA_OPENSSL			openssl
-	set_var EASYRSA_PKI				"$EASYRSA/pki"
+	set_var EASYRSA_PKI				"/tmp/easyrsa3/pki"
 	set_var EASYRSA_DN				cn_only
 	set_var EASYRSA_REQ_COUNTRY		"US"
 	set_var EASYRSA_REQ_PROVINCE	"California"
--- a/easyrsa3/openssl-easyrsa.cnf
+++ b/easyrsa3/openssl-easyrsa.cnf
@@ -104,8 +104,8 @@ emailAddress			= Email Address
 emailAddress_default		= $ENV::EASYRSA_REQ_EMAIL
 emailAddress_max		= 64
 
-serialNumber		= Serial-number (eg, device serial-number)
-serialNumber_default	= $ENV::EASYRSA_REQ_SERIAL
+#serialNumber		= Serial-number (eg, device serial-number)
+#serialNumber_default	= $ENV::EASYRSA_REQ_SERIAL
 
 ####################################################################
 # Easy-RSA cert extension handling
