--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -185,18 +185,33 @@ set(CMAKE_EXPORT_COMPILE_COMMANDS ON)
 
 if (ANDROID)
     set(ANDROID_MY_LIBS ${PROJECT_SOURCE_DIR}/trojan-plus-android-libs) 
+
+    set(BOOST_VERSION_HPP ${ANDROID_MY_LIBS}/include/boost/version.hpp)
+    if(EXISTS ${BOOST_VERSION_HPP})
+        file(STRINGS ${BOOST_VERSION_HPP} BOOST_VERSION_LINE REGEX "^#define BOOST_VERSION ")
+        string(REGEX REPLACE "^#define BOOST_VERSION ([0-9]+)$" "\\1" BOOST_VERSION ${BOOST_VERSION_LINE})
+    endif()
+
     set(ANDROID_MY_LIBS_LIBRARIES 
         ${ANDROID_MY_LIBS}/lib/${ANDROID_ABI}/libssl.a 
         ${ANDROID_MY_LIBS}/lib/${ANDROID_ABI}/libcrypto.a
-        ${ANDROID_MY_LIBS}/lib/${ANDROID_ABI}/libboost_system.a
         ${ANDROID_MY_LIBS}/lib/${ANDROID_ABI}/libboost_program_options.a)
 
+    if (Boost_MAJOR_VERSION LESS_EQUAL 1 AND Boost_MINOR_VERSION LESS 89)
+        list(APPEND ANDROID_MY_LIBS_LIBRARIES ${ANDROID_MY_LIBS}/lib/${ANDROID_ABI}/libboost_system.a)
+    endif()
+
     set(OPENSSL_VERSION 1.1.1)
 
     include_directories(${ANDROID_MY_LIBS}/include)
     target_link_libraries(trojan ${ANDROID_MY_LIBS_LIBRARIES})
 else()
-    find_package(Boost 1.66.0 REQUIRED COMPONENTS system program_options)
+    find_package(Boost 1.66.0 REQUIRED)
+    if (Boost_MAJOR_VERSION LESS_EQUAL 1 AND Boost_MINOR_VERSION LESS 89)
+        find_package(Boost 1.66.0 REQUIRED COMPONENTS system program_options)
+    else()
+        find_package(Boost 1.66.0 REQUIRED COMPONENTS program_options)
+    endif()
     include_directories(${Boost_INCLUDE_DIR})
     target_link_libraries(trojan ${Boost_LIBRARIES})
     if(MSVC)
